web/api/v1/api.go

```diff
@@ -805,6 +808,93 @@ func rulesAlertsToAPIAlerts(rulesAlerts []*rules.Alert) []*Alert {
 	return apiAlerts
 }
 
+type metricsWithMetadata map[string]metadataSet
+
+type metadataSet map[metadata]struct{}
+
+func (ms metadataSet) MarshalJSON() ([]byte, error) {
+	var res bytes.Buffer
+	json := jsoniter.ConfigCompatibleWithStandardLibrary
+	encoder := json.NewEncoder(&res)
+
+	res.WriteByte('[')
+	for metadata := range ms {
+		if err := encoder.Encode(metadata); err != nil {
+			return nil, err
+		}
+	}
+	res.WriteByte(']')
+
+	return res.Bytes(), nil
+}
+
+type metadata struct {
+	Type textparse.MetricType `json:"type"`
+	Help string               `json:"help"`
+	Unit string               `json:"unit"`
+}
+
+func (api *API) metricMetadata(r *http.Request) apiFuncResult {
+	var metrics metricsWithMetadata
+
+	limit := -1
+	if s := r.FormValue("limit"); s != "" {
+		var err error
+		if limit, err = strconv.Atoi(s); err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, errors.New("limit must be a number")}, nil, nil}
+		}
+	}
+
+	metrics = make(metricsWithMetadata)
```

Comment:
Why not this this all in one line? Why does this type even need to live beyond this function?